core/query/delateralize.ml

```diff
@@ -0,0 +1,247 @@
+    | Q.Variant (label, v) -> Q.Variant (label, srec v)
+    | Q.Apply (f, xs) -> Q.Apply (srec f, List.map srec xs)
+    | Q.For (_, gs, os, u) ->
+        (* XXX: assuming fresh x!*)
+        let gs' = List.map (fun (v,g) -> (v, srec g)) gs in
+        let os' = List.map srec os in
+        let u' = srec u in
+        Q.For (None, gs', os', u')
+    | Q.If (c, t, e) ->
+        Q.If (srec c, srec t, srec e)
+    | Q.Case (v, cases, default) ->
+        let v' = srec v in
+        let cases' = StringMap.map (fun (v,q) -> (v,srec q)) cases in
+        let default' = default >>= fun d -> Some (fst d, srec (snd d)) in
+        Q.Case (v', cases', default')
+    | Q.Dedup v -> Q.Dedup (srec v)
+    | Q.Prom v -> Q.Prom (srec v)
+    | Q.Closure (c, closure_env) ->
+        let cenv = MixingQuery.Lang.bind closure_env (x,u) in
+        Q.Closure (c, cenv)
+    | v -> v
+
+(* C(q1, x.q2) := for x :- q1, #y :- q2 do {(x,#y)}
+   also returns the fieltypes of the graph *)
+let graph_query (q1,ty1) x (q2,ty2) =
+    let y = Var.fresh_raw_var () in
+    let p = Q.flattened_pair (Q.Var (x,ty1)) (Q.Var (y,ty2)) in
+    let ftys = Q.flattened_pair_ft (Q.Var (x,ty1)) (Q.Var (y,ty2)) in
+    Q.For (None, [(x, q1); (y, q2)], [], Q.Singleton p), ftys
+
+(* DELATERALIZING REWRITE for iota
+     for gs, y :- I(q3) do q1     -- s.t. x :- q2 in gs
+    ~> for gs, p :- I(C(Dq2,x.q3)) where x = p.1 do ([y]q1) p.2 *)
+let prom_delateralize gs q1 x (q2,ty2) y (q3,ty3) =
+    let p = Var.fresh_raw_var () in
+    let graph, ftys = graph_query (Q.Dedup q2,ty2) x (q3,ty3) in
+    let vp = Q.Var (p,Types.make_record_type ftys) in
+    let vx = Q.Var (x,ty2) in
+    let eq_test a b = Q.Apply (Q.Primitive "==", [a;b]) in
+    let and_query a b = Q.Apply (Q.Primitive "&&", [a;b]) in
+    (* eta-expanded vx == p.1, with record flattening *)
+    (* UNSMART? there may be a better way of doing this, because normalization already pushes equality of records
+       to a conjunction of equalities over their fields; however, here we are using a flattened version of the
+       record p, so extracting p.1 really amounts to building a new record; maybe it wouldn't be much smarter, after all *)
+    let eq_query =
+        StringMap.fold
+            (fun f _ acc -> and_query acc (eq_test (Q.Project (vx, f)) (Q.Project (vp, Q.flatfield "1" f))))
+            (Q.recdty_field_types ty2)
+            (Q.Constant (Constant.Bool true))
+    in
+    (* eta-expanded p.2, with record flattening *)
+    let rp =
+        Q.Record
+            (StringMap.fold
+                (fun f _ acc -> StringMap.add f (Q.Project (vp, Q.flatfield "2" f)) acc)
+                (Q.recdty_field_types ty3)
+                StringMap.empty)
+    in
+    let q1_rp = subst q1 y rp
+    in
+    Q.For (None,
+        gs @ [(p, Q.Prom graph)],
+        [],
+        Q.If (eq_query, q1_rp, Q.nil))
+
+(* Returns (Some ty) if v occurs free with type ty, None otherwise *)
+let occurs_free (v : Var.var) =
```

Comment:
Also seems like it could go in mixingQuery

Reply:
Moved to `QueryLang`